### PR TITLE
Keyboard navigation improvements

### PR DESCRIPTION
**Pull request by web-padawan**
Connected to #9

### CODE REVIEW

File: src/vaadin-menu-bar-buttons-mixin.html [hunk 1]
```diff
@@ -99,6 +99,7 @@
     }
 
     _onItemClick(e) {
+      e.preventDefault();
```
Comment (pekam) [inline src/vaadin-menu-bar-buttons-mixin.html]: Why is this needed?

Comment (web-padawan) [reply src/vaadin-menu-bar-buttons-mixin.html]: That is related to the issue described in https://github.com/vaadin/vaadin-menu-bar/issues/9#issuecomment-477059015

> ↓ (arrow down) should not scroll window when opening sub-menu

By default, arrow down scrolls the page, which shouldn't happen with menu-bar.

**Comment by web-padawan** [comment]
@pekam I can also reproduce the issue you mentioned with a context-menu.
Submitted as vaadin/vaadin-context-menu#230

**Comment by web-padawan** [comment]
Moving this PR back to in progress, as it makes sense to cover all the keys at once.
I will split those into separate commits to ensure the resulting code is easy to review.

File: src/vaadin-menu-bar-buttons-mixin.html (hunk 1 above)
Comment (pekam) [reply src/vaadin-menu-bar-buttons-mixin.html]: I see. To improve code readability the function name could be changed (I had the impression that this was called only on click events based on the name). I don't have good suggestions though... :D Not blocking.

**Comment by pekam** [comment]
> Moving this PR back to in progress, as it makes sense to cover all the keys at once.
I will split those into separate commits to ensure the resulting code is easy to review.

However you like to do it. You can also merge this one already if you wish.

**Comment by web-padawan** [comment]
More fixes to come (I will update the list when pushing)

- [x] Up Arrow - Opens submenu and moves focus to last item in the submenu.
- [ ] Left Arrow - Close submenu, focus previous menu bar item, open submenu for it (but not focus)
- [ ] Right Arrow - Close submenu, focus next menu bar item, open submenu for it (but not focus)

**Comment by web-padawan** [comment]
Keyboard navigation and roles to be implemented in according to WAI ARIA guidelines:
https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-1/menubar-1.html

### Menubar

<table aria-labelledby="kbd1_label kbd_label" class="def">
      <thead>
        <tr>
          <th>Key</th>
          <th>Function</th>
          <th>Done</th>
        </tr>
      </thead>
      <tbody>
        <tr data-test-id="menubar-space-or-enter">
          <th>
            <kbd>Space</kbd><br><kbd>Enter</kbd>
          </th>
          <td>Opens submenu and moves focus to first item in the submenu.</td>
          <td>Yes</td>
        </tr>
        <tr data-test-id="menubar-right-arrow">
          <th>
            <kbd>Right Arrow</kbd>
          </th>
          <td>
            <ul>
              <li>Moves focus to the next item in the menubar.</li>
              <li>If focus is on the last item, moves focus to the first item.</li>
            </ul>
          </td>
          <td>Yes</td>
        </tr>
        <tr data-test-id="menubar-left-arrow">
          <th>
            <kbd>Left Arrow</kbd>
          </th>
          <td>
            <ul>
              <li>Moves focus to the previous item in the menubar.</li>
              <li>If focus is on the first item, moves focus to the last item.</li>
            </ul>
          </td>
          <td>Yes</td>
        </tr>
        <tr data-test-id="menubar-down-arrow">
          <th>
            <kbd>Down Arrow</kbd>
          </th>
          <td>Opens submenu and moves focus to first item in the submenu.</td>
          <td>Yes</td>
        </tr>
        <tr data-test-id="menubar-up-arrow">
          <th>
            <kbd>Up Arrow</kbd>
          </th>
          <td>Opens submenu and moves focus to last item in the submenu.</td>
          <td></td>
        </tr>
        <tr data-test-id="menubar-home">
          <th>
            <kbd>Home</kbd>
          </th>
          <td>Moves focus to first item in the menubar.</td>
         <td>Yes</td>
        </tr>
        <tr data-test-id="menubar-end">
          <th>
            <kbd>End</kbd>
          </th>
          <td>Moves focus to last item in the menubar.</td>
         <td>Yes</td>
        </tr>
        <tr data-test-id="menubar-character">
          <th>
            <kbd>Character</kbd>
          </th>
          <td>
            <ul>
              <li>Moves focus to next item in the menubar having a name that starts with the typed character.</li>
              <li>If none of the items have a name starting with the typed character, focus does not move.</li>
            </ul>
          </td>
         <td>No<sup>1</sup></td>
        </tr>
      </tbody>
    </table>

### Submenu

<table aria-labelledby="kbd2_label kbd_label" class="def">
      <thead>
        <tr>
          <th>Key</th>
          <th>Function</th>
          <th>Done</th>
        </tr>
      </thead>
      <tbody>
        <tr data-test-id="submenu-space-or-enter">
          <th>
            <kbd>Space</kbd><br><kbd>Enter</kbd>
          </th>
          <td>
            <ul>
              <li>Activates menu item.</li>
            </ul>
          </td>
         <td>Yes</td>
        </tr>
        <tr data-test-id="submenu-escape">
          <th>
            <kbd>Escape</kbd>
          </th>
          <td>
            <ul>
              <li>Closes submenu.</li>
              <li>Moves focus to parent menubar item.</li>
            </ul>
          </td>
           <td>Yes</td>
        </tr>
        <tr data-test-id="submenu-right-arrow">
          <th>
            <kbd>Right Arrow</kbd>
          </th>
          <td>
            <ul>
              <li>If focus is on an item with a submenu, opens the submenu and places focus on the first item. </li>
              <li>If focus is on an item that does not have a submenu:
                <ul>
                  <li>Closes submenu.</li>
                  <li>Moves focus to next item in the menubar.</li>
                  <li>Opens submenu of newly focused menubar item, keeping focus on that parent menubar item.</li>
                </ul>
              </li>
            </ul>
          </td>
           <td>Yes</td>
        </tr>
        <tr data-test-id="submenu-left-arrow">
          <th>
            <kbd>Left Arrow</kbd>
          </th>
          <td>
            <ul>
              <li>Closes submenu and moves focus to parent menu item.</li>
              <li>If parent menu item is in the menubar, also:
                <ul>
                  <li>moves focus to previous item in the menubar.</li>
                  <li>Opens submenu of newly focused menubar item, keeping focus on that parent menubar item.</li>
                </ul>
              </li>
            </ul>
          </td>
           <td>Yes</td>
        </tr>
        <tr data-test-id="submenu-down-arrow">
          <th>
            <kbd>Down Arrow</kbd>
          </th>
          <td>
            <ul>
              <li>Moves focus to the next item in the submenu.</li>
              <li>If focus is on the last item, moves focus to the first item.</li>
            </ul>
          </td>
           <td>Yes</td>
        </tr>
        <tr data-test-id="submenu-up-arrow">
          <th>
            <kbd>Up Arrow</kbd>
          </th>
          <td>
            <ul>
              <li>Moves focus to previous item in the submenu.</li>
              <li>If focus is on the first item, moves focus to the last item.</li>
            </ul>
          </td>
           <td>1. Yes 2. No<sup>2</sup></td>
        </tr>
        <tr data-test-id="submenu-home">
          <th>
            <kbd>Home</kbd>
          </th>
          <td>Moves focus to the first item in the submenu.</td>
           <td>Yes</td>
        </tr>
        <tr data-test-id="submenu-end">
          <th>
            <kbd>End</kbd>
          </th>
          <td>Moves focus to the last item in the submenu.</td>
           <td>Yes</td>
        </tr>
        <tr data-test-id="submenu-character">
          <th>
            <kbd>Character</kbd>
          </th>
          <td>
            <ul>
              <li>Moves focus to the next item having a name that starts with the typed character.</li>
              <li>If none of the items have a name starting with the typed character, focus does not move.</li>
            </ul>
          </td>
           <td>Yes</td>
        </tr>
      </tbody>
    </table>

<sup>1</sup> See #31 
<sup>2</sup> We close the submenu instead.

**Comment by web-padawan** [comment]
@pekam nice catch, currently menu is closed because of outside click listener. Will fix.

But in order to open submenu on <kbd>Enter</kbd> or <kbd>Space</kbd>, we also need to update context-menu so it would work accordingly (it only accepts <kbd>Arrow Right</kbd> currently).

File: test/sub-menu.html [hunk 2]
```diff
@@ -88,21 +89,106 @@
...
+        await onceOpened(subMenu);
         item = subMenu.$.overlay.querySelector('vaadin-context-menu-item');
+        expect(item).to.be.ok;
+        await nextRender(subMenu);
         arrowUp(item);
         await nextRender(subMenu);
         expect(subMenu.opened).to.be.false;
       });
 
+      (isIOS ? it.skip : it)('should switch menubar item and open submenu on arrow left', async() => {
```
Comment (pekam) [inline test/sub-menu.html]: When pressing <kbd>Left</kbd> on a sub-menu, if the left button does not have children, the focus remains on the original button, and the focus-ring is lost entirely. I'd expect the focus to move to the left button regardless if it has children or not.

If I press <kbd>Right</kbd> on a non-parent sub-menu item, and the next button does not have children, it still gets the focus, like I'd expect.

(added comment on this line instead of a general comment just to get threads)

**Comment by web-padawan** [comment]
> Currently it closes the menu.

Pushed the fix for closing menu on parent item Enter / Space.

File: test/sub-menu.html (hunk 2 above)
Comment (web-padawan) [reply test/sub-menu.html]: Fixed. Had to use `stopImmediatePropagation()` for <kbd>Left</kbd> to stop handling in context-menu.

File: src/vaadin-menu-bar-interactions-mixin.html [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    ready() {
+      super.ready();
+
+      this.addEventListener('keydown', e => this._onKeydown(e));
+      this.addEventListener('focusin', e => this._onFocusin(e));
+
+      this._subMenu.addEventListener('item-selected', this.__onItemSelected.bind(this));
+      this._subMenu.addEventListener('close-all-menus', this.__onEscapeClose.bind(this));
```
Comment (pekam) [inline src/vaadin-menu-bar-interactions-mixin.html]: The menu should also close by Esc key when the focus is on a button.
To reproduce:
1. Open a sub-menu
2. Hit the Left arrow key to open the previous sub-menu while keeping the focus on the button instead of an item
3. Hit the Esc key and notice that the menu is not closed

In the w3-example the menu closes in a similar case.

File: src/vaadin-menu-bar-interactions-mixin.html [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    __fireItemSelected(value) {
+      this.dispatchEvent(new CustomEvent('item-selected', {detail: {value}}));
+    }
+
+    __onButtonClick(e) {
+      e.stopPropagation();
+      const button = this._getButtonFromEvent(e);
+      this.__openSubMenu(button);
```
Comment (pekam) [inline src/vaadin-menu-bar-interactions-mixin.html]: Needs a falsy-check for `button`. Currently, if I click the menu-bar without hitting any button, an error is thrown at line 252.

File: src/vaadin-menu-bar-interactions-mixin.html (hunk 3 above)
Comment (web-padawan) [reply src/vaadin-menu-bar-interactions-mixin.html]: Good catch. Will fix.

File: src/vaadin-menu-bar-interactions-mixin.html [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+    _onMouseOver(e) {
+      const button = Array.from(e.composedPath()).filter(el => el.item)[0];
+      if (button && button !== this._expandedButton && (this.openOnHover || (button.item.children && this._subMenu.opened))) {
+        this.__openSubMenu(button);
+      }
+    }
+
+    __onContextMenuKeydown(e) {
+      const item = e.target;
```
Comment (pekam) [inline src/vaadin-menu-bar-interactions-mixin.html]: It's possible to have something else than a `vaadin-context-menu-item` here.
When clicking at the edge of the overlay, the whole overlay is focused and that will be the event target. Then the page can be scrolled with arrow keys, and Up arrow throws an error.

File: src/vaadin-menu-bar-interactions-mixin.html (hunk 4 above)
Comment (web-padawan) [reply src/vaadin-menu-bar-interactions-mixin.html]: Fixed.

File: src/vaadin-menu-bar-interactions-mixin.html (hunk 3 above)
Comment (web-padawan) [reply src/vaadin-menu-bar-interactions-mixin.html]: Fixed.

File: src/vaadin-menu-bar-interactions-mixin.html [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+        const button = this.shadowRoot.activeElement;
+        if (button && button.item && button.item.children) {
+          this.__openSubMenu(button, {keepFocus: true});
+        }
+      }
+    }
+
+    __outsideClickListener(e) {
+      const path = e.composedPath();
+      if (path.indexOf(this) === -1 && path.indexOf(this._subMenu.$.overlay) === -1) {
```
Comment (pekam) [inline src/vaadin-menu-bar-interactions-mixin.html]: Should check if it targets any buttons, instead of checking if it targets the menu-bar.

Currently eg. in the demo, it is really easy to make an outside click without closing the overlay. The buttons consume only a portion of the menu-bar width, so the empty part of the menu-bar is easy to hit.

Comment (web-padawan) [reply src/vaadin-menu-bar-interactions-mixin.html]: Fixed.

File: src/vaadin-menu-bar-interactions-mixin.html (hunk 5 above)
Comment (web-padawan) [reply src/vaadin-menu-bar-interactions-mixin.html]: Fixed. At least the error is no longer thrown, I'm not sure what should we do about focusing the overlay though. It might be better to discuss in person and possibly fix later.

Comment (pekam) [reply src/vaadin-menu-bar-interactions-mixin.html]: Well that's a really special case, and it's not easily reproduced by accident, so I think this is good enough for now.

File: src/vaadin-menu-bar-interactions-mixin.html [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+    _focusFirstItem() {
+      const list = this._subMenu.$.overlay.firstElementChild;
+      list.focus();
+    }
+
+    _focusLastItem() {
+      const list = this._subMenu.$.overlay.firstElementChild;
+      const item = list.items[list.items.length - 1];
+      item.focus();
```
Comment (pekam) [inline src/vaadin-menu-bar-interactions-mixin.html]: A really special case again, but a falsy check for `item` would avoid an error here, when the user has only components (non-vaadin-item) inside the overlay. I know that at least one Flow user has used context-menu with just a combo-box in the overlay.

File: test/sub-menu.html [hunk 8]
```diff
@@ -88,19 +94,144 @@
...
+        arrowLeft(item);
+        await nextRender(subMenu);
         item = subMenu.$.overlay.querySelector('vaadin-context-menu-item');
-        arrowUp(item);
+        const spy = sinon.spy(item, 'focus');
+        arrowDown(buttons[2]);
+        expect(spy).to.be.calledOnce;
+      });
+
+      (isIOS ? it.skip : it)('should focus last item on arrow down after opened on arrow left', async() => {
```
Comment (pekam) [inline test/sub-menu.html]: typo: arrow up

Comment (web-padawan) [reply test/sub-menu.html]: Fixed.

File: test/sub-menu.html [hunk 9]
```diff
@@ -189,6 +331,17 @@
           await nextRender(subMenu);
           expect(buttons[0].hasAttribute('expanded')).to.be.false;
         });
+
+        it('should remove expanded attribute and restore focus when sub-menu closed on Esc', async() => {
+          arrowDown(buttons[0]);
+          await nextRender(subMenu);
+          expect(buttons[0].hasAttribute('expanded')).to.be.true;
+
+          MockInteractions.pressAndReleaseKeyOn(subMenu.$.overlay, 27, [], 'Escape');
```
Comment (pekam) [inline test/sub-menu.html]: I guess this could use the `esc` helper function? not blocking

Comment (web-padawan) [reply test/sub-menu.html]: Thanks, done.

File: src/vaadin-menu-bar-interactions-mixin.html (hunk 7 above)
Comment (web-padawan) [reply src/vaadin-menu-bar-interactions-mixin.html]: Thanks, fixed.
